### PR TITLE
refactor: cross-server post and friends api

### PR DESCRIPTION
**Pull request by ExiaSR**
Get /author/posts and /posts no longer break if one of the
foreign server is broken

/posts endpoint now also query from foreign servers

Added 'required_trailing_slash' in Server model,
to handle some weird foreign servers

Reformat foreign user profile string presentation

## Check list

- [x] My branch is up to date with `master`. [How?](https://github.com/ExiaSR/hyperion/wiki/How-to-collaborate)
- [x] I passed all integration tests in Travis CI.
- [x] Now it is a good time to ask for review.

### CODE REVIEW

File: hyperion/views/friend_views.py [hunk 1]
```diff
@@ -189,14 +190,9 @@ def friend_request(request):
...
                     }
-                    # print(friend_request_body,'htz')
-                    resp = requests.post(
-                        url=friend_host_name + "/api/friendrequest",
-                        json=friend_request_body,
-                        auth=(
-                            friend_profile.host.foreign_db_username,
-                            friend_profile.host.foreign_db_password,
-                        ),
+                    foreign_server = Server.objects.get(url=friend_host_name)
```
Comment (Raymundo1) [inline hyperion/views/friend_views.py]: I don't think server has `url` attribute
```Python
foreign_server = friend_profile.host
```
不是很明白ForeignServerHttpUtils.post里面的server: Server这种用法，为啥这里的server会有endpoint

Comment (ExiaSR) [reply hyperion/views/friend_views.py]: @Raymundo1 
在这个PR里面加入的新东西 #139 

https://github.com/ExiaSR/hyperion/blob/master/hyperion/models/server.py#L16

url是server的hostname
endpoint是实际server的api endpoint
这两个field是为了方便可以直接从Server里面通过URL来filter，然后抓取endpoint和credential

**Comment by Raymundo1** [comment]
[Some Place also need changed](https://github.com/ExiaSR/hyperion/blob/341b1b767373b08493be264ac38ed9f074c5bfbd/hyperion/views/friend_views.py#L302-L309)
不好意思，不太知道怎么直接显示code,我直接给的link
这里麻烦可能也要改一下
```Python
# in this case
foreign_server = friend_request_obj.from_profile.host
```

File: hyperion/views/friend_views.py (hunk 1 above)
Comment (Raymundo1) [reply hyperion/views/friend_views.py]: 关于ForeignServerHttpUtils.post(server: Server, ...)，这种用法是可以的对吗

Comment (ExiaSR) [reply hyperion/views/friend_views.py]: @Raymundo1 这个叫python的typed parameters
https://docs.python.org/3/library/typing.html

方便IDE做autocomplete，然后一眼看也知道传进来的是什么东东。

**Pull request merged**